include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1723 @@
+  void ref_store(const variable_t &ref, const variable_t &rgn,
+                 const variable_or_constant_t &val) override {
+    crab::CrabStats::count(domain_name() + ".count.ref_store");
+    crab::ScopedCrabStats __st__(domain_name() + ".ref_store");
+
+    // Perform a store.
+    // ref_store(ref, rgn, val);
+    // Definition: assign abstract value in val to rgn
+    //  rgn: a region variable used in object domain
+    //  res: a register variable used in base domain
+    // The transformer is sound for ref_store required the followings:
+    //  1. The type of region variable is consistent with the type of register
+    //  E.g. if region represents integer, the register must be an integer type
+    //  2. The reference ref is not evaluated as null pointer.
+    //  3. The number of references implicitly implies:
+    //  a. if is one, it means abstract object is singleton,
+    //     perform strong update: update the contents of an object in
+    //     abstraction is the same as the concrete
+    //  b. if is more than one, it means abstract object is a summarized object,
+    //     perform a weak update: update the contents of an object in
+    //     abstraction is an over-approximation here: the region is updated for
+    //     new val as combined the old properties (through join op)
+    //
+    // For a. the property of this abstract object is in the base domain:
+    //    perform assign in base dom, except if val is not in base domain
+    //
+    // For b. the object domain of the abstract object is either:
+    //     object domain is Top (i.e. the object domain is not found in odi map)
+    //  or object domain exists
+    //
+    //  if object domain is Top, create a new object domain and set into odi map
+    //  if object domain exists, create a new one by copying old one
+    //
+    //  post condition: the odi map is updated, the base domain is not changed
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+    ERROR_IF_ARRAY_REGION(rgn, __LINE__);
+    ERROR_IF_NOT_REF(ref, __LINE__);
+    if ((rgn.get_type().is_bool_region() && !val.get_type().is_bool()) ||
+        (rgn.get_type().is_integer_region() && !val.get_type().is_integer()) ||
+        (rgn.get_type().is_real_region() && !val.get_type().is_real()) ||
+        (rgn.get_type().is_reference_region() &&
+         !val.get_type().is_reference())) {
+      CRAB_ERROR(domain_name(), "::ref_store: type of value ", val, " (",
+                 val.get_type(), ") is not compatible with region ", rgn, " (",
+                 rgn.get_type(), ")");
+    }
+
+    if (is_bottom()) {
+      return;
+    }
+
+    // for now skip analysis for unknown region
+    if (is_unknown_region(rgn)) {
+      return;
+    }
+
+    if (auto id_opt = get_obj_id(rgn)) {
+      // After instrisic calls, the object id must exist for each region if the
+      // region belongs to an abstract object
+
+      // retrieve an abstract object info
+      auto obj_info = m_obj_info_env.at(*id_opt);
```

Comment:
As a general comment, note that `separate_domain` already provides `const Value* find(const Key &k)`. So in cases like this you can avoid copies